File: Year 3/Parallel and Distributed Programming/ConvolutionMatrix Java/src/Convolutions/ConvolutionTaskCols.java
package Convolutions;

import Singleton.Convolution;

public class ConvolutionTaskCols extends Thread {
    private final int[][] inputMatrix;
    private final int[][] convolutionMatrix;
    private final int[][] resultMatrix;
    private final int startCol;
    private final int endCol;
    private final Convolution convolution = Convolution.getInstance();

    public ConvolutionTaskCols(int[][] inputMatrix, int[][] convolutionMatrix, int[][] resultMatrix, int startCol, int endCol) {
        this.inputMatrix = inputMatrix;
        this.convolutionMatrix = convolutionMatrix;
        this.resultMatrix = resultMatrix;
        this.startCol = startCol;
        this.endCol = endCol;
    }

    @Override
    public void run() {
        int N = inputMatrix.length;
        int M = inputMatrix[0].length;
        for (int j = startCol; j < endCol && j < M; j++) {
            for (int i = 0; i < N; i++) {
                resultMatrix[i][j] = convolution.applyConvolution(inputMatrix, convolutionMatrix, i, j);
            }
        }
    }
}

File: Year 3/Parallel and Distributed Programming/Laboratory 1/ConvolutionMatrix C++/ConvolutionMatrix C++ Parallel/Source Files/Convolutions/ConvolutionTaskCols.cpp
#include "../../Header Files/Convolutions/ConvolutionTaskCols.h"

ConvolutionTaskCols::ConvolutionTaskCols(const vector<vector<int>>& inputMatrix, const vector<vector<int>>& convolutionMatrix,
	vector<vector<int>>& resultMatrix, int startCol, int endCol)
	: inputMatrix(inputMatrix), convolutionMatrix(convolutionMatrix), resultMatrix(resultMatrix), startCol(startCol), endCol(endCol) {
	convolution = Convolution::getInstance();
}

void ConvolutionTaskCols::run() {
	int N = sizeof(inputMatrix.size());
	int M = sizeof(inputMatrix[0].size());
	for (int j = startCol; j < endCol && j < M; j++) {
		for (int i = 0; i < N; i++) {
			resultMatrix[i][j] = convolution->applyConvolution(inputMatrix, convolutionMatrix, i, j);
		}
	}
}
